### ISSUE TITLE
False positives

### ISSUE THREAD

**Issue by certainlyakey** [ismay/stylelint-no-unsupported-browser-features#45]
Thanks for the plugin!

I'm not sure if it's a bug with stylelint-no-unsupported-browser-features or with one of the underlying packages, but when a preprocessor (SASS) function containing `calc` word in its name is used the `calc` feature is reported as being used.

For example, with `letter-spacing:rem-calc(1.8px);` this is reported: 

```
Unexpected browser feature "calc" is not supported by Blackberry Browser 7, plugin/no-unsupported-browser-features Opera Mobile 12,12.1 and only partially supported by Android Browser 4.4,4.4.3-4.4.4
```

The browserlist is 

```
"browserslist": [
    "last 3 versions",
    "not ie_mob > 0",
    "not ie <= 10",
    "not OperaMini all"
  ]
```

**Comment by ismay** [comment]
Hmm ok. Difficult to judge without seeing your code. Could you create a repo with the minimum code that reproduces the bug?

**Comment by ismay** [comment]
I'll close this for now. Let me know if you keep encountering this issue and in that case please provide the code to reproduce the issue.

**Comment by certainlyakey** [comment]
OK, sorry, forgot to post a code example. 

```
.c-excerpt {
  margin-bottom:rem-calc(52px); 
  hyphens:auto;
  text-align:center; 
}
```

in a `_textcontent.scss` is reported as 

```
scss/_textcontent.scss
  3:3   ✖  Unexpected browser feature "calc" is not supported by Opera Mobile 12.1 and     plugin/no-unsupported-browser-features
           only partially supported by Android Browser 4.4.3-4.4.4
```

**Comment by ismay** [comment]
@certainlyakey Thanks for the reply, but what I meant was a complete example, including your stylelint setup. This only shows me the markup, but it doesn't tell me how you configured stylelint.

**Comment by certainlyakey** [comment]
OK, sure! Here's the full example:

`.stylelintrc.json`:

```
{
  "extends": "stylelint-config-recommended-scss",
  "plugins": [
    "stylelint-no-unsupported-browser-features"
  ],
  "rules": {
    "plugin/no-unsupported-browser-features": true
  }, 
  "ignoreFiles": ["scss/autogenerated/**/*.scss", "scss/overrides/**/*.scss"]
}
```

`package.json`:
```
{
  "name": "...",
  "version": "1.0.0",
  "dependencies": {
    "SassyCast": "github:HugoGiraudel/SassyCast",
    "babel-preset-es2015": "^6.5.0",
    "babelify": "^7.2.0",
    "browserify": "^13.0.0",
    "dotenv": "^4.0.0",
    "focus-trap": "^2.3.0",
    "gulp": "^3.9.1",
    "gulp-autoprefixer": "^3.1.0",
    "gulp-concat": "^2.6.0",
    "gulp-cssnano": "^2.1.1",
    "gulp-gettext": "^0.3.0",
    "gulp-jshint": "^2.0.0",
    "gulp-modernizr": "^1.0.0-alpha",
    "gulp-notify": "^2.2.0",
    "gulp-plumber": "^1.1.0",
    "gulp-sass": "^3.1.0",
    "gulp-sass-variables": "^1.1.0",
    "gulp-sourcemaps": "^2.1.1",
    "gulp-stylelint": "^5.0.0",
    "gulp-svg-sprite": "^1.3.1",
    "gulp-uglify": "^1.5.2",
    "gulp-util": "^3.0.7",
    "jquery": "^3.0.0",
    "jshint": "^2.9.2",
    "jshint-stylish": "^2.0.1",
    "normalize.css": "^3.0.1",
    "sass-module-importer": "^1.4.0",
    "sass-mq": "^3.3.2",
    "sass-rem": "^2.0.0",
    "stylelint-config-recommended-scss": "^2.0.0",
    "stylelint-no-px": "^0.12.0",
    "stylelint-scss": "^2.0.0",
    "susy": "^2.2.12",
    "vinyl-buffer": "^1.0.0",
    "vinyl-source-stream": "^1.1.0",
},
  "browserslist": [
    "last 2 versions",
    "not ie_mob > 0",
    "not BlackBerry > 0",
    "not ie <= 10",
    "not OperaMini all"
  ]
}
```

`_teasers.scss`
```
.class {
  width:rem-calc(20px);
}
```

Terminal: 

```
scss/components/teasers/_teasers.scss
 2:3  ✖  Unexpected browser feature "calc" is not supported by Opera Mobile 12.1 and only  plugin/no-unsupported-browser-features
         partially supported by Android Browser 4.4.3-4.4.4
```

**Comment by ismay** [comment]
Hmm ok. Yeah no idea where this is going wrong. Would have to check that. But it seems like a valid bug. Thanks for reporting it!

**Comment by certainlyakey** [comment]
Cool, thanks!

**Comment by ismay** [comment]
Same thing seems to happen with a variable named transition, which is recognised as a regular css transition.

**Comment by ismay** [comment]
#49 seems related. Seems like variable and functions names that match css properties are being interpreted as those properties.

**Comment by snaumov** [comment]
the problems seem to be caused by the fact that at the moment all of the non-CSS declarations are sent as is to the doiuse plugin, and doiuse simply looks up if the declaration (like in my case) or declaration value (like in @certainlyakey case) is contained in the unsupported feature name

`function isFoundIn(str) {
  str = stripUrls(str);
  return function find(searchfor) {
    if (searchfor instanceof RegExp) return searchfor.test(str);else if (_.isFunction(searchfor)) return searchfor(str);else return str && str.indexOf(searchfor) >= 0;
  };
}`

Do not know yet if it's possible to send already transpiled syntax to doiuse

**Comment by ismay** [comment]
Took a look at it, and it is a doiuse issue. This plugin merely [passes the css to doiuse](https://github.com/ismay/stylelint-no-unsupported-browser-features/blob/master/index.js#L35) and logs the warnings doiuse returns. Please open this issue at the doiuse repo: https://github.com/anandthakker/doiuse.

FYI, I used to help maintain it, but stepped down as I don't have the time. If you want this to be fixed feel free to volunteer as a maintainer there, very much needed!

**Comment by snaumov** [comment]
@ismay it seems that the problem is in between - while debugging I see that your plugin sends all variables declarations and properties in non-standard SCSS syntax _as is_ to doiuse, and doiuse only supports CSS syntax (undercover it simply checks if the incoming properties names contain non-supported features, i.e. "filter" is in "$background-filter").

Do you maybe know if, just guessing, Stylelint can return parsed CSS to your plugin so it can be sent over to doiuse in proper form?

**Comment by ismay** [comment]
Hmm ok. Which version of stylelint did you test with? I might’ve encountered this earlier. Wonder if it’s the plugin’s responsibility to handle various syntaxes, because that seems weird.

**Comment by snaumov** [comment]
Tested with the latest stylelint 8.4.0, and 8.1.0 as well.

**Comment by snaumov** [comment]
so here `doiuse(doiuseOptions).postcss(root, doiuseResult)` root will contain all kind of SCSS, i.e.: `root.nodes[0].prop` equal to $filter-background, which later doiuse tries to parse, etc.

**Comment by ismay** [comment]
Ok, so I had a similar issue before and discussed it here: https://github.com/stylelint/stylelint/issues/2904

So it seems that there are two problems:

1. scss functions get passed to doiuse (which it cannot handle)
2. css variable declarations get passed to doiuse (which it also doesn't handle correctly)

But doiuse should [ignore nodes that it cannot or shouldn't handle](https://github.com/anandthakker/doiuse/blob/7383ca5ae833738613e362d26f951a478df9712c/src/detect-feature-use.js#L116). An scss function is apparently recognised as a valid node type, as well as a css variable declaration.

Maybe it's useful to ignore declarations that might cause false positives. But I think it's better to handle this in doiuse. Because otherwise we're just smoothing over stuff that should actually be fixed in doiuse.

**Comment by ismay** [comment]
Opened to prevent similar issues from being opened, but I'm still not convinced that this is something that should be handled in this plugin.

**Comment by snaumov** [comment]
As doiuse currently supports only CSS syntax and expects correct postcss AST tree to be received, I still not sure if it's doiuse responsibility to handle current case correctly.

Doiuse will skip unknown node types, but current AST tree received from stylelint contains legit nodes, with types _decl, _rule.

I see three ways to solve the problem, each on different level of tools involved :)

1)Return "correct" AST from stylelint - didn't find a way to do this yet, and based on [discussion you mentioned](https://github.com/stylelint/stylelint/issues/2904), seems to be that the stylelint will return AST with non-standard syntax, so plugin creators using utils as `isStandardSyntax` can proceed.

2)On plugin level edit the AST tree and send over only supported properties to doiuse.

3)On level of doiuse correctly check property names [here](https://github.com/anandthakker/doiuse/blob/7383ca5ae833738613e362d26f951a478df9712c/src/detect-feature-use.js#L12-L19), and in particular replace `return str && str.indexOf(searchfor) >= 0` with strict equality check `return str && str === searchfor`.
I don't know what was the purpose of partial, non-strict comparison, but maybe there is one?

**Comment by ismay** [comment]
> As doiuse currently supports only CSS syntax and expects correct postcss AST tree to be received, I still not sure if it's doiuse responsibility to handle current case correctly.

As for not passing it scss, I agree. However, it should handle css variables properly. So we might need some input from the stylelint team, because changing the AST will also require us to deal with a bunch of other stuff, like sourcemaps (to map errors to the right column, line), etc. Plus, apparently stylelint doesn't directly export its `isStandard...` utils, so I'm not sure what their recommended approach is.

> I don't know what was the purpose of partial, non-strict comparison, but maybe there is one?

Might be because of vendor prefixing.

**Comment by ismay** [comment]
So I opened this to see what the stylelint team recommends, as I suspect that we're not the first dealing with this: https://github.com/stylelint/stylelint/issues/2904#issuecomment-352799949

**Comment by jeddy3** [comment]
> Return "correct" AST from stylelint - didn't find a way to do this yet, and based on discussion you mentioned, seems to be that the stylelint will return AST with non-standard syntax

That's right. stylelint returns a PostCSS AST with non-standard constructs shoved into one of the available node types (`rule`, `atrule`, `decl`, and `comment`). For example, dollar variables (e.g. `$var: my-value`) are `decl` nodes within the tree (e.g. `{ nodes: { type: decl, prop: $var, value: "my-value } }`. The [various PostCSS parsers](https://github.com/postcss/postcss#syntaxes) (e.g. `postcss-scss`, `postcss-less`, `postcss-sass`, `sugarss` etc) do this shoving for their respective syntaxes.

stylelint plugin authors can then choose what they want to do with the non-standard constructs in the tree. For example, they could choose to:

1. do nothing about it and just document what syntaxes the plugin is compatible with
2. add code to ignore non-standard constructs if they are only targetting standard ones
3. specifically target some non-standard construct (like `stylelint-scss` does with [its rules](https://github.com/kristerkari/stylelint-scss/tree/master/src/rules))

> As doiuse currently supports only CSS syntax and expects correct postcss AST tree to be received

It's perfectly valid for doiuse (and this stylelint plugin) to only support standard CSS syntax. Just document this. Supporting (even just ignoring) non-standard syntax in doiuse could be [a lot of hard work](http://davidtheclark.com/lessons-i-may-have-learned/#non-standard-syntaxes-are-a-huge-obstacle-to-css-tooling) and might not be worth the effort. 

---

I agree with @ismay that any changes should happen in doiuse. This stylelint plugin should just be a thin wrapper for that PostCSS plugin.

With that in mind, I think you're on the right path for adding CSS Variables support to doiuse. Using a strict comparison for `decl.prop` should do it. You can use [`postcss.vendor.unprefixed()`](http://api.postcss.org/vendor.html#.unprefixed) to handle vendor prefixes. (Incidentally, as a side-effect, this means dollar variables will also be correctly ignored.)

If someone thinks it's worth the effort and wants to add support for ignoring non-standard functions, like `rem-calc()`, to doiuse I suggest reaching for "sub" parsers like [postcss-value-parser](https://github.com/TrySound/postcss-value-parser). This will grant finer-grained (and strict) comparisons (on things like function names), rather than operating on the whole `value` property of the `decl` nodes.

**Comment by snaumov** [comment]
> With that in mind, I think you're on the right path for adding CSS Variables support to doiuse. Using a strict comparison for decl.prop should do it. You can use postcss.vendor.unprefixed() to handle vendor prefixes. (Incidentally, as a side-effect, this means dollar variables will also be correctly ignored.)

Was trying for some time to implement that, but realized, that the loose comparison of `decl.prop` is used in doiuse not only for vendor prefixing elimination, but also to match the whole set of properties, having i.e. `background` in their [names](https://github.com/anandthakker/doiuse/blob/master/data/features.js#L198), and loose comparison of `decl.value` is also needed since we can have something like `red background-gradient(blah)` as a property value and we need to look up for `background-gradient`. So the strict comparison will not play nicely with this design. 

So to offload everything to doiuse I think we need to implement there something similar to stylelint isStandartSyntax* utils, and skip non-standart nodes. 
For `rem-calc()` case we can use the mentioned parser.

What'd you say?

**Comment by ismay** [comment]
@jeddy3 Thank you for your feedback. Ok, so what I'll do for now is follow your recommendations and document that this plugin is only compatible with standard css syntax. We can always think about adding support for non-standard syntax, but then at the moment it's at least clear what is and isn't supported.

Also, I'll open an issue with doiuse for the false positives when using css variables. I think that should cover everything we've discussed in this issue. Thanks for the help @jeddy3 and @snaumov. Feel free to keep the discussion going and let me know if I've forgotten to address anything.
